### ISSUE TITLE
Please add support for custom name display formats

### ISSUE THREAD

**Issue by syominsergey** [gramps-project/gramps-web#444]
Hi!

In my culture preferrable order of names looks like this:
```
<given name> <patronymic surname> <surname>
```
Or another order also possible:
```
<surname> <given name> <patronymic surname>
```

But in GrampsWeb another order of names is widely used:
```
<given name> <surname> <patronymic surname>
```

Or another possible order of such set of names in GrampsWeb:
```
<surname> <patronymic surname> <given name>
```

All these order existing in GrampsWeb are looks very strange and unusual in my culture.

Lets look at some examples.

List of persons (`https://<domain>/people`):
![image](https://github.com/gramps-project/gramps-web/assets/28873347/dcf9e50c-c6ab-48c9-967e-c25206c1c289)

Info about concrete person (`https://<domain>/person/<person_id>`):
![image](https://github.com/gramps-project/gramps-web/assets/28873347/ed89331c-49e5-4e3e-8b45-c2f42d6fabb0)

Ancestor tree and other types of trees (`https://<domain>/tree`):
![image](https://github.com/gramps-project/gramps-web/assets/28873347/93b8f5f2-5e3e-4339-8408-005c026354a5)

Maybe some other views exists with same order of names unusual for my culture.

Desktop Gramps application has elegant way to solve this problem. It has setting for for custom name display format:
https://gramps-project.org/wiki/index.php/Gramps_5.2_Wiki_Manual_-_Settings#Display_Name_Editor

I can set it to desired value and after it order of names looks natively to my culture in all views I have found in desktop app. For example some screenshots from Desktop Gramps 5.2.2 after setting of custom name display format (sorry for screenshots from desktop app - they are in translation to my language, I can't find howto change translation to English in desktop app).

List of persons:
![image](https://github.com/gramps-project/gramps-web/assets/28873347/b8338385-49e7-4ba5-aa7c-410fe4e9d792)

Info about concrete person:
![image](https://github.com/gramps-project/gramps-web/assets/28873347/9eaedabe-51d4-4fef-b253-d7d2b5b09daf)

Tree:
![image](https://github.com/gramps-project/gramps-web/assets/28873347/8ab26f73-87e4-44b9-a319-bbae0f5ff234)

Please, implement ability to change default order of names in all views in GrampsWeb according to setting of custom name display format in GrampsWeb, just like works in desktop Gramps.

P.s.

Version of GrampsWeb where I created screenshots:
```
Gramps 5.2.2
Gramps Web API 2.3.0
Gramps Web Frontend 24.5.0
Gramps QL 0.3.0
locale: en
multi-tree: false
task queue: false
```

P.p.s. I have discussed this feature request at [gramps discourse group](https://gramps.discourse.group/t/is-it-possible-to-use-custom-name-display-formats-in-grampsweb/5495).
Also I found another feature request is very close to current: https://github.com/gramps-project/gramps-web/issues/411.
It is about ability to change names order from `Surname, FirstName` to `FirstName Surname` in tree, graph & Chart view.
I think, it is a part of current more generic feature request.
So, after discussion in [gramps discourse group](https://gramps.discourse.group/t/is-it-possible-to-use-custom-name-display-formats-in-grampsweb/5495) I decided to open this new feature request.

**Comment by andye2801** [comment]
When this is implemented would this allow us to change the display to show married names by default?

**Comment by DavidMStraub** [comment]
No, I don't think so. Gramps treats married names as alternative names and Gramps Web uses the primary name everywhere, because otherwise it's difficult to determine which altnernative name to pick in which case.

**Comment by andye2801** [comment]
Is it possible to select which name is the primary name manually?

**Comment by DavidMStraub** [comment]
Not right now. It's part of this feature request. But that one is simple to implement.

**Comment by nikallass** [comment]
Also it would be very good to have an ability to show person in this format:
<given name> <patronymic surname> <married surname> (<maiden surname>)

Ex:
`Natalia Alexandrovna Smirnova (Ivanova)`

**Comment by DavidMStraub** [comment]
> Is it possible to select which name is the primary name manually?

I ended up creating a separate issue for that one after all, as it's a simple UI change, while the name display formats are more work.

https://github.com/gramps-project/gramps-web/issues/562

**Comment by DavidMStraub** [comment]
I see three options for implementing custom name display formats:

- Adding a custom name format query parameter to the person endpoint, adding the formatted name to the person profile, and using the person profile for all displayed names in the frontend
- Adding a new endpoint where a raw name object can be posted and a formatted name is returned
- Reimplementing `gramps.gen.display.name` in Javascript

I haven't investigated this in detail enough to judge which is the best road.

I think this feature is very well motivated, but I'm personally not likely to work on this anytime soon for purely egoistic reasons (I don't need it). If anyone wants to take it on, I can provide more detailed instructions.

**Comment by vktimofeev** [comment]
Hi everybody,

I would try to implement this feature request.

I investigated the code of both frontend and Web API, and I found the following:

The "strange and unusual" rendering of a person name, as @syominsergey qualified it, having the form:

<given name> <surname> <patronymic surname>

takes place because the **name_given** and **name_surname** of the "**profile**" are generated by  the **get_person_profile_for_object()** method defined in the **gramps_webapi/api/resources/util.py** file (line number 455 in the version I'm working with) as follows:

 ```
profile = {
 <skipped lines>
        "name_given": name_display.display_given(person),
        "name_surname": person.primary_name.get_surname(),
<skipped lines>
}

```
get_surname() is a method implemented in the gen.lib.surnamebase and inherited by the **Name** class of GRAMPS backend. It is proceeding by joining all the surnames contained within a Name object altogether, including their prefixes and connectors into a single string which is returned.  This behavior is hardcoded and cannot be altered by any configuration parameters for now.

But patronymic names are treated as surnames in GRAMPS, and this is exactly why the "strange and unusual" form of the surname is rendered.

Taking into account the following option mentioned above

> * Adding a custom name format query parameter to the person endpoint, adding the formatted name to the person profile, and using the person profile for all displayed names in the frontend

I wonder whether it is feasible that both **name_given** and **name_surname** of the "**profile**" were configured via custom format query parameter. I can explain why both may require custom format: sometimes it is desirable to separate patronymic "surname" from the primary surname  - this is the case of   the http(s)://_servername_/people page and some others, but in the same time it is desirable to join the patronymic "surname" with the given name in the second position.

 I've already made a "quick and dirty" hack in my personal installation of the Gramps Web by modifying the same gramps_webapi/api/resources/util.py as follows.

```
profile = {
 <skipped lines>
        "name_given": name_display.format_str(person.get_primary_name(), "%f %1y")
        "name_surname": name_display.primary_surname(person.get_primary_name()),
<skipped lines>
}
```

This code works fine for me on all the pages that I viewed after the modification, but of course it cannot be used to replace the current code in the **master** branch.

So I would like to ask whether such solution is acceptable at all, and if it is so, to demand more detailed instructions.

I suggest that **format_str** method of the **NameDisplay** class from the gramps.gen.display.name can be used for the name and surname formatting  on the API side, and I wonder why it might be necessary to reimplement this code in JavaScript?

**Comment by DavidMStraub** [comment]
Thanks for volunteering to tackle this @vktimofeev!

First, regarding your last question - no, I'm not saying it's necessary to reimplement it in Javascript. In my previous post, the three bullet points are options, not steps. So, your suggestion to modify the person profile corresponds to my first bullet point and is probably the best approach.

Rather than modifying the existing `name_given` and `name_surname` keys, I think it's better to introduce a new `name_display` key in the person profile and allow arbitrary name formats, just as in Gramps desktop. This will also allow users to choose between given name first and surname first formats.

The second step would be to add the analogous frontend code in the user settings, where users will be able to change the name format to their liking.

Since this is the frontend repo, I suggest to use the current issue for this second step and I'll create a new one in the backend.

Let me know what you think.

**Comment by greewoo** [comment]
As mentioned in #1167, in genealogy, writing surnames in ALL CAPS (e.g., John SMITH) [...].

In my opinion, a viable solution would be to allow users to customize the display format for their surname in the settings (normal, uppercase).
